### PR TITLE
[FLINK-5179] Correct TM MetricRegistry and TmMG lifecycle

### PR DESCRIPTION
**Pull request by zentol**
This PR corrects the life-cycle of the MetricRegistry and TaskManagerMetricGroup of the TaskManager.

The MetricRegistry is no longer shutdown when the TaskManager disassociates from the JobManager; the registry, and by extension the reporters, will now live and be active as long as the TaskManager is running.

The TaskManagerMetricGroup is now properly closed when disassociating from the JobManager.

### CODE REVIEW

**Comment by rmetzger** [comment]
I wonder if there is an easy way to add a test case for this?
Maybe you could add a check to one of the HA tests?

**Comment by zentol** [comment]
That should be possible; I'll look into it.

**Comment by zentol** [comment]
@rmetzger I've added a test case.

**Comment by uce** [comment]
This has been marked as a blocker for 1.1.4. Is there a chance that we will get this in today?

**Comment by zentol** [comment]
Someone has to review the test and then we're good to go IMO.

**Comment by zentol** [comment]
Actually, i don't know whether @rmetzger did review the functional changes; but that's only a single line.

File: flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala [hunk 1]
```diff
@@ -1046,11 +1046,10 @@ class TaskManager(
       network.getKvStateRegistry.unregisterListener()
     }
     
-    // failsafe shutdown of the metrics registry
     try {
-      metricsRegistry.shutdown()
+      taskManagerMetricGroup.close()
```
Comment (uce) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala]: When is this shut down then?

Comment (zentol) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala]: When the taskmanager is shutdown, in postStop().

Comment (zentol) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala]: That was already the case before this PR.

**Comment by uce** [comment]
The changes and the test look good to me. While waiting for Travis, I will test this with an HA cluster. We can merge this if everything succeeds.

**Comment by uce** [comment]
I tested this, but could not reproduce the reported erroneous behaviour. After the task manager reconnects to a new job manager, the metrics are still being reported. I tested this with JMX and Log4j reporters. The changes look reasonable to me though. @rmetzger @zentol Should we merge this for 1.1.4?

**Comment by uce** [comment]
Can you open a PR with a backport of this for 1.1.4 please?

**Comment by zentol** [comment]
Sorry, while back-porting i realized this issue does not affect 1.1.4. I've adjusted the JIRA as such.

**Comment by uce** [comment]
OK that also explains why it didn't occur for me with the current release-1.1 branch. ;)

**Comment by uce** [comment]
Tested this vs. the current master and it works as expected. In the current master, the metrics are not updated after the TM disassociates from the JM. With this PR the reporter is closed, restarted  and metrics are being reported again.

In general, I'm wondering whether we should introduce an explicit suspended state where the metrics are still available and only reset after recovery. I was testing this with JMX and wondered whether it can be problematic for users if the reporter are down during JM recovery.

**Comment by zentol** [comment]
The reporter isn't closed and restarted; instead all metrics are being de-registered and then re-registered. This is necessary since the ID of the TM changes upon re-association with the JM.

It would make sense to only un-register the metrics after we associate with a new JM; this should reduce the downtime of taskmanager metrics.

**Pull request closed without merging**